lib/Sema/MiscDiagnostics.cpp

```diff
@@ -927,7 +927,7 @@ static void diagSyntacticUseRestrictions(const Expr *E, const DeclContext *DC,
       }
       
       // Unchecked casting to a subclass is better done by unsafeDowncast.
-      if (fromTy->isBindableToSuperclassOf(toTy)) {
+      if ((!fromTy->hasArchetype() && fromTy->isExactSuperclassOf(toTy))) {
```

Comment:
Maybe we should just use `TypeChecker::isSubtypeOf(fromTy, toTy, ...)` here then?

Reply:
Either way works really, if you want to change DiagnoseWalker in this PR you can do that separately.